<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание,-цели-и-задачи-исследования" data-toc-modified-id="Описание,-цели-и-задачи-исследования-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание, цели и задачи исследования</a></span></li><li><span><a href="#Знакомство-с-данными" data-toc-modified-id="Знакомство-с-данными-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Знакомство с данными</a></span></li><li><span><a href="#Посчитаем-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посчитаем-сколько-книг-вышло-после-1-января-2000-года-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Посчитаем сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Для-каждой-книги-посчитаем-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитаем-количество-обзоров-и-среднюю-оценку-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Для каждой книги посчитаем количество обзоров и среднюю оценку</a></span></li><li><span><a href="#Определим-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-мы-исключим-из-анализа-брошюры" data-toc-modified-id="Определим-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-мы-исключим-из-анализа-брошюры-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры</a></span></li><li><span><a href="#Определим-автора-с-самой-высокой-средней-оценкой-книг-—-учитем-только-книги-с-50-и-более-оценками" data-toc-modified-id="Определим-автора-с-самой-высокой-средней-оценкой-книг-—-учитем-только-книги-с-50-и-более-оценками-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Определим автора с самой высокой средней оценкой книг — учитем только книги с 50 и более оценками</a></span></li><li><span><a href="#Посчитаем-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок." data-toc-modified-id="Посчитаем-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок.-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок.</a></span></li><li><span><a href="#выводы" data-toc-modified-id="выводы-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>выводы</a></span></li></ul></div>

# Исседование сервиса для чтения книг по подписке

## Описание, цели и задачи исследования

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Наша задача — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

Ход работы:
1. Для начала подключить библиотеки, чтобы можно использовать язык запросов SQL и загрузим базу данных
1. Познакомимся с данными и посчитаем сколько строк в каждой таблице.
2. В рамках исседования отетим на следующие вопросы:
- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.
4. Сформулируем выводы

## Знакомство с данными

In [1]:
# импортируем библиотеки
import pandas as pd
import sqlalchemy as sa

In [2]:
# устанавливаем параметры
db_config = {
'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db' # название базы данных
}

In [3]:
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

In [4]:
engine = sa.create_engine(connection_string, connect_args={'sslmode':'require'})

In [5]:
# чтобы выполнить SQL-запрос, пишем функцию с использованием Pandas
def get_sql_data(query:str, engine:sa.engine.base.Engine=engine) -> pd.DataFrame:
    '''Открываем соединение, получаем данные из sql, закрываем соединение'''
    with engine.connect() as con:
        return pd.read_sql(sql=sa.text(query), con = con)

Выведем по одной строчке для каждой таблицы, а также посчитаем общее количество строк

In [6]:
# формируем запрос и выводим данные
query = '''
SELECT *
FROM books 
LIMIT 1
'''
get_sql_data(query)

book_id  author_id         title  num_pages publication_date  publisher_id
0        1        546  'Salem's Lot        594       2005-11-01            93

- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

In [7]:
query = '''
SELECT COUNT(*) AS total_rows
FROM books 

'''
get_sql_data(query)

total_rows
0        1000

Всего у нас 1000 книг

In [8]:
query = '''
SELECT * 
FROM authors 
LIMIT 1
'''
get_sql_data(query)

author_id      author
0          1  A.S. Byatt

- author_id — идентификатор автора;
- author — имя автора.

In [9]:
query = '''
SELECT COUNT(*) AS total_rows
FROM authors 

'''
get_sql_data(query)

total_rows
0         636

Авторов 636

In [10]:
query = '''
SELECT * 
FROM publishers 
LIMIT 1
'''
get_sql_data(query)

publisher_id publisher
0             1       Ace

- publisher_id — идентификатор издательства;
- publisher — название издательства;

In [11]:
query = '''
SELECT COUNT(*) AS total_rows
FROM publishers  

'''
get_sql_data(query)

total_rows
0         340

Издателей 340

In [12]:
query = '''
SELECT * 
FROM ratings 
LIMIT 1
'''
get_sql_data(query)

rating_id  book_id    username  rating
0          1        1  ryanfranco       4

- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

In [13]:
query = '''
SELECT COUNT(*) AS total_rows
FROM ratings  

'''
get_sql_data(query)

total_rows
0        6456

оценок 6456

In [14]:
query = '''
SELECT * 
FROM reviews
LIMIT 1
'''
get_sql_data(query)

review_id  book_id      username  \
0          1        1  brandtandrea   

                                                text  
0  Mention society tell send professor analysis. ...

- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя пользователя, написавшего обзор;
- text — текст обзора.

In [15]:
query = '''
SELECT COUNT(*) AS total_rows
FROM reviews 

'''
get_sql_data(query)

total_rows
0        2793

Вывод по разделу:
1. Всего в нашем сервисе 1000 книг. Есть столбцы с названием, датой выспуска и количеством сраниц, а также уникальные идентификаторы книги, атора и издателя, являющиеся ключами.
2. 636 автора, что меньше книг. В столбце с названием содержится полное имя и фамилия автора. Таблицы связаны  ключем с идентификатором автора.
3. 340 издателей таблица содержит название издателя и ключ с индентификатором издателя.
4. 6456 оцени книги с юзернеймом пользоватля, оставившего оценку. Таблица содержит идентификатор книги.
5. 2793 отзывов. Пользователи чаще оставляют оценку, чем пишут отызв, как и должно быть. Таблица содержит идентификтор книги.

База данных соответствует обозначенной структуре.

## Посчитаем сколько книг вышло после 1 января 2000 года

In [16]:
query = '''
SELECT COUNT(book_id)
FROM books
WHERE  publication_date::date  > '2000-01-01'

'''
get_sql_data(query)

count
0    819

Всего книг у на 1000 из них 819 вышли после 2000 года. Значит у нас книги современных издательств!

## Для каждой книги посчитаем количество обзоров и среднюю оценку

In [17]:
query = '''
WITH ar AS (SELECT
    b.book_id as id,
    b.title AS title,
    AVG (r.rating) OVER (PARTITION  BY b.book_id) AS avg_rating 
FROM books b
LEFT JOIN
ratings r ON b.book_id=r.book_id),
tx AS (SELECT book_id as id,
         COUNT (review_id) OVER (PARTITION  BY  book_id) AS cnt_review
FROM reviews)
SELECT DISTINCT ar.title,
        ar.avg_rating,
        tx.cnt_review
FROM ar
LEFT JOIN
tx ON ar.id=tx.id

ORDER BY avg_rating DESC, cnt_review DESC

'''
get_sql_data(query)

title  avg_rating  cnt_review
0     The Cat in the Hat and Other Dr. Seuss Favorites        5.00         NaN
1                        A Dirty Job (Grim Reaper  #1)        5.00         4.0
2         Moneyball: The Art of Winning an Unfair Game        5.00         3.0
3              School's Out—Forever (Maximum Ride  #2)        5.00         3.0
4                A Fistful of Charms (The Hollows  #4)        5.00         2.0
..                                                 ...         ...         ...
995  The World Is Flat: A Brief History of the Twen...        2.25         3.0
996                                      Drowning Ruth        2.00         3.0
997                  His Excellency: George Washington        2.00         2.0
998                                              Junky        2.00         2.0
999                               Harvesting the Heart        1.50         2.0

[1000 rows x 3 columns]

Мы посчитали средние оценки и количестве обзоров на книгу. Оценки начинаются от 1.5 до 5. Книга "грязная работа" получила все "пятерки" и на нее написано 4 обзора. Так же можно увидеть, что не на все книги написаны обзоры.

## Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры

In [18]:
query = '''
SELECT DISTINCT p.publisher AS publisher,
    COUNT (b.book_id) OVER (PARTITION  BY p.publisher) AS number_books 
FROM books b
JOIN
publishers p ON b.publisher_id=p.publisher_id
WHERE b.num_pages > 50


ORDER BY number_books DESC

LIMIT 1
'''
get_sql_data(query)

publisher  number_books
0  Penguin Books            42

Издатель "Penguin Books" издал больше всего книг, в которых больше 50 страниц - 42 штуки.

## Определим автора с самой высокой средней оценкой книг — учитем только книги с 50 и более оценками

In [22]:
query = '''
WITH au AS (
    SELECT b.book_id AS book_id,
           a.author AS author
    FROM books b
    JOIN
    authors a ON b.author_id = a.author_id),

b AS (SELECT b.book_id AS book_id
    FROM books b
    JOIN
    ratings r ON b.book_id=r.book_id
    GROUP BY b.book_id
    HAVING COUNT(r.rating)>=50),

aub AS (SELECT b.book_id AS book_id,
       au.author AS author
FROM b
LEFT JOIN 
au ON b.book_id=au.book_id)

SELECT DISTINCT a.author AS author,
    AVG (r.rating) OVER (PARTITION  BY a.author) AS avg_rating

FROM aub a
LEFT JOIN
ratings r ON a.book_id = r.book_id

ORDER BY avg_rating DESC

LIMIT 1

'''
get_sql_data(query)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.287097

Самый рейтинговый автор J.K. Rowling/Mary GrandPré

## Посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [20]:
query = '''
SELECT ROUND(SUM(cnt_review)/COUNT(*)) AS avg_review
FROM(
SELECT  DISTINCT r.username AS name,
        COUNT(DISTINCT w.review_id) AS cnt_review
FROM ratings r
LEFT JOIN
reviews w ON r.username = w.username
GROUP BY r.username
HAVING COUNT(DISTINCT r.rating_id)>48
) AS A
'''
get_sql_data(query)

avg_review
0        24.0

В среднем клиенты поставившие более 48 оценок делают 24 обзора. Клиенты, которые часто оценивают фильмы очень активны в написании обзоров

## выводы
1. В нашем новом сервисе преимущественно книги современных издательств. 
2. Мы сделали список самых рейтинговых книг и количество обзоров на них. Не на все даже самые рейтинговые книги написаны обзоры. Минимальная оценка книги 1,5 максимальная 5.
3. Издатель "Penguin Books" издал больше всего книг, в которых больше 50 страниц - 42 штуки.
4. Самый рейтинговый автор J.K. Rowling/Mary GrandPré
5. Активные пользователи в среднем оставляют 24 отзыва.
Рекомендации
- Продвигать наш сервис среди людей, которые ищут современных писателей.
- Предлагать новым пользователям, рейтинговые книги с большим количеством отзывов. Например,  Dirty Job, Moneyball: The Art of Winning an Unfair Game, School's Out—Forever
- С популярным издательством Penguin Books, можно заключить контракт с экслюзивными правами на книге по подписке, чтобы создать конкурентные преимущества
- автора J.K. Rowling/Mary GrandPré стоит назначить амбассадором сервиса. Его очень любят наши пользователи.
- Активным пользователям, которые оставляют больше 20 отзывов, можно делать скидки на подписку. Они будут рекламировать наши книги.